presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java

```diff
@@ -327,4 +320,109 @@ private static int calculateMaxFill(int hashSize)
         checkArgument(hashSize > maxFill, "hashSize must be larger than maxFill");
         return maxFill;
     }
+
+    private class AddPageWork
+            implements Work<Void>
+    {
+        private final Block block;
+
+        private boolean finished;
+        private int lastPosition;
+
+        public AddPageWork(Block block)
+        {
+            this.block = requireNonNull(block, "block is null");
+        }
+
+        @Override
+        public boolean process()
+        {
+            if (finished) {
+                return true;
+            }
+
+            if (needRehash() && !tryRehash()) {
+                return false;
+            }
+
+            int positionCount = block.getPositionCount();
+            verify(lastPosition < positionCount, "position count out of bound");
+            while (lastPosition < positionCount) {
+                // get the group for the current row
+                putIfAbsent(lastPosition, block);
+                lastPosition++;
+
+                if (needRehash()) {
+                    // no need to try rehashing; putIfAbsent should have done that
```

Comment:
if should have done that, then why not `checkState(! needRehash())` ?

Reply:
same in `Multichannel...`

Reply:
I should just remove the comment; I was trying to say something like:

"we don't need the pattern:"
```
if (needRehash()) {
    tryRehash();
    break;
}
```
because this has been done in `putIfAbsent()`